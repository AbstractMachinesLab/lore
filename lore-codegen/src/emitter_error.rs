use miette::Diagnostic;
use std::path::PathBuf;
use thiserror::Error;

#[derive(Diagnostic, Error, Debug, PartialEq, Eq)]
#[error("Error emitting file {filename:?}")]
#[diagnostic(code(lore::codegen::emitter), url(docsrs))]
pub struct EmitterError {
    filename: PathBuf,
}
