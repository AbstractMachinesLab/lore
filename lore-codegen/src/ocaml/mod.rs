pub mod ocaml_ast;
pub mod ocaml_emitter;

pub use ocaml_ast::*;
pub use ocaml_emitter::*;
