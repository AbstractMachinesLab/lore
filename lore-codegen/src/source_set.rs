use miette::Diagnostic;
use std::path::PathBuf;
use thiserror::Error;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Source {
    name: PathBuf,
    contents: String,
}

#[derive(Error, Debug, Diagnostic)]
#[error("Error writting file {filename:?}")]
#[diagnostic(code(lore::codegen::source_set), url(docsrs))]
pub struct SourceError {
    filename: PathBuf,

    #[source]
    error: std::io::Error,
}

impl Source {
    pub fn new(name: PathBuf, contents: String) -> Source {
        Source { name, contents }
    }

    pub fn write(&self, prefix: &PathBuf) -> Result<(), SourceError> {
        let path = prefix.join(self.name.clone());
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent).map_err(|error| SourceError {
                filename: path.clone(),
                error,
            })?;
        }
        std::fs::write(&path, &self.contents).map_err(|error| SourceError {
            filename: path,
            error,
        })
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct SourceSet {
    sources: Vec<Source>,
}

impl SourceSet {
    pub fn empty() -> SourceSet {
        SourceSet::default()
    }

    pub fn from_sources(sources: Vec<Source>) -> SourceSet {
        SourceSet { sources }
    }

    pub fn sources(&self) -> &Vec<Source> {
        &self.sources
    }
}
