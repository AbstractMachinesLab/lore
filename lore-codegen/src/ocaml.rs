use crate::emitter_error::*;
use crate::source_set::*;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CamlName {
    parts: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CamlModule {
    name: CamlName,
    structure: Vec<CamlValue>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum CamlValue {
    Module(CamlModule),
    Type { name: String },
}

impl CamlName {
    pub fn from_kind_name(name: &lore_ast::Name) -> CamlName {
        let mut parts = vec![];
        for part in name.to_string().to_lowercase().replace("/", ":").split(":") {
            parts.push(part.to_string())
        }

        if let Some(first) = parts[0].get_mut(0..1) {
            first.make_ascii_uppercase();
        };

        CamlName { parts }
    }

    pub fn to_filename(&self) -> String {
        self.parts.join("_")
    }
}

impl Into<Source> for CamlModule {
    fn into(self) -> Source {
        let filename = format!("{}.mli", self.name.to_filename());
        let contents = format!("{}", self);
        Source::new(filename.into(), contents)
    }
}

impl std::fmt::Display for CamlName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
        write!(f, "{}", self.parts[0])?;
        for part in self.parts[1..].iter() {
            write!(f, "_{}", part)?;
        }
        Ok(())
    }
}

impl std::fmt::Display for CamlModule {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
        for item in &self.structure {
            write!(f, "{}\n", item)?;
        }
        Ok(())
    }
}

impl std::fmt::Display for CamlValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
        match self {
            CamlValue::Module(m) => write!(f, "{}", m),

            CamlValue::Type { name } => write!(f, "type {}", name),
        }
    }
}

#[derive(Default)]
pub struct OCamlEmitter {}

// impl LoreEmitter<Document> for OCamlEmitter {
impl OCamlEmitter {
    pub fn new() -> OCamlEmitter {
        OCamlEmitter::default()
    }

    pub fn translate(&self, store: &lore_store::Store) -> Result<SourceSet, EmitterError> {
        let mut sources = vec![];

        for kind in store.kinds() {
            let module = CamlModule {
                name: CamlName::from_kind_name(&kind.name),
                structure: vec![CamlValue::Type {
                    name: "t".to_string(),
                }],
            };
            sources.push(module.into())
        }

        Ok(SourceSet::from_sources(sources))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::*;

    macro_rules! test {
        ($name:ident, $src:expr) => {
            #[test]
            fn $name() {
                let mut store = lore_store::Store::new();
                let store = store.add_from_string($src).unwrap();
                let emitter = OCamlEmitter::new();
                let caml_value = emitter.translate(&store).unwrap();
                let snapshot = format!(
                    r#"
input:
    {}

output:

{:#?}
"#,
                    $src, caml_value
                );
                assert_snapshot!(snapshot)
            }
        };
    }

    test!(
        kind_to_type,
        r#"

use spotify:ontology:v2022/Artist as Artist
use spotify:ontology:v2022/Album as Album
use spotify:ontology:v2022/Track as Track
use spotify:ontology:v2022/Name as Name
use spotify:ontology:v2022/hasOne as hasOne
use spotify:ontology:v2022/isListedIn as isListedIn

kind Artist

kind Album

kind Track

attr Name

rel Album hasOne Name

rel Track isListedIn Album

        "#
    );
}
