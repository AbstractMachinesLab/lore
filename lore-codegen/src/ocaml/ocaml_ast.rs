use crate::source_set::*;
use std::fmt::{Display, Error, Formatter};
use std::path::PathBuf;

///////////////////////////////////////////////////////////////////////////////////////////////////
///
/// An OCaml Module
///

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct CamlModuleName(String);

impl CamlModuleName {
    pub fn from_name(name: &lore_ast::Name) -> CamlModuleName {
        let mut parts = vec![];
        for part in name.to_string().replace("/", ":").split(":") {
            parts.push(part.to_string())
        }

        let mut name = parts.join("_").to_lowercase();

        if let Some(first) = name.get_mut(0..1) {
            first.make_ascii_uppercase();
        };

        CamlModuleName(name)
    }
}

impl Display for CamlModuleName {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        write!(f, "{}", self.0)
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CamlModule {
    name: CamlModuleName,
    filename: PathBuf,
    structure: Vec<CamlValue>,
}

impl Into<Source> for CamlModule {
    fn into(self) -> Source {
        Source::new(self.filename.clone(), format!("{}", self))
    }
}

impl Display for CamlModule {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        for item in &self.structure {
            write!(f, "{}", item)?;
            writeln!(f)?;
        }
        Ok(())
    }
}

impl CamlModule {
    pub fn new(name: CamlModuleName) -> CamlModule {
        let filename = PathBuf::from(format!("{}.mli", name.to_string().to_lowercase()));
        CamlModule {
            name,
            filename,
            structure: vec![],
        }
    }

    pub fn with_structure(self, structure: Vec<CamlValue>) -> CamlModule {
        CamlModule { structure, ..self }
    }
}

///////////////////////////////////////////////////////////////////////////////////////////////////
///
/// An OCaml Type
///

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct CamlFieldName(String);

impl Display for CamlFieldName {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        write!(f, "{}", self.0)
    }
}

impl CamlFieldName {
    pub fn from_name(name: &lore_ast::Name) -> CamlFieldName {
        let mut parts = vec![];
        for part in name.to_string().replace("/", ":").split(":") {
            parts.push(part.to_string())
        }
        let mut name = parts.join("_");
        if let Some(first) = name.get_mut(0..1) {
            first.make_ascii_lowercase();
        };
        CamlFieldName(name)
    }
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct CamlRecord {
    fields: Vec<(CamlFieldName, CamlType)>,
}

impl Display for CamlRecord {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        let mut fields = self.fields.clone();
        fields.sort_by(|a, b| a.cmp(b));

        write!(f, "{{ {}: {}", fields[0].0, fields[0].1)?;
        for (name, typ) in fields[1..].iter() {
            write!(f, "\n; {}: {}", name, typ)?;
        }
        write!(f, " }}")
    }
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum CamlVariant {
    Constructor(Vec<CamlType>),
    InlineRecord(CamlRecord),
}

impl Display for CamlVariant {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        match self {
            CamlVariant::InlineRecord(r) => {
                write!(f, " of {}", r)
            }

            CamlVariant::Constructor(cs) => {
                if !cs.is_empty() {
                    write!(f, " of ")?;
                    write!(f, "{}", cs[0])?;
                    for c in cs[1..].iter() {
                        write!(f, "* {}", c)?;
                    }
                };
                Ok(())
            }
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum CamlType {
    Reference {
        module_path: CamlModuleName,
        type_name: String,
    },
    Abstract(String),
    Variant {
        name: String,
        variants: Vec<CamlVariant>,
    },
    Record {
        name: String,
        record: CamlRecord,
    },
}

impl CamlType {
    pub fn reference(module_path: CamlModuleName, type_name: String) -> CamlType {
        CamlType::Reference {
            module_path,
            type_name,
        }
    }

    pub fn abstract_type(name: String) -> CamlType {
        CamlType::Abstract(name)
    }

    pub fn record(name: String, fields: Vec<(CamlFieldName, CamlType)>) -> CamlType {
        CamlType::Record {
            name,
            record: CamlRecord { fields },
        }
    }
}

impl Display for CamlType {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        match self {
            CamlType::Reference {
                module_path,
                type_name,
            } => write!(f, "{}.{}", module_path, type_name),
            CamlType::Abstract(name) => write!(f, "type {}", name),
            CamlType::Record { name, record } => {
                write!(f, "type {} = {} \n", name, record)
            }
            CamlType::Variant { name, variants } => {
                write!(f, "type {} = \n", name)?;
                for v in variants {
                    write!(f, "| {}", v)?;
                }
                Ok(())
            }
        }
    }
}

///////////////////////////////////////////////////////////////////////////////////////////////////
///
/// An OCaml Value
///

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum CamlValue {
    Module(CamlModule),
    Type(CamlType),
}

impl Display for CamlValue {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        match self {
            CamlValue::Module(m) => write!(f, "{}", m),
            CamlValue::Type(t) => write!(f, "{}", t),
        }
    }
}
