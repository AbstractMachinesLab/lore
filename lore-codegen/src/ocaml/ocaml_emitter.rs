use crate::emitter_error::*;
use crate::ocaml::ocaml_ast::*;
use crate::source_set::*;

#[derive(Default)]
pub struct OCamlEmitter {}

// impl LoreEmitter<Document> for OCamlEmitter {
impl OCamlEmitter {
    pub fn new() -> OCamlEmitter {
        OCamlEmitter::default()
    }

    pub fn translate(&self, store: &lore_store::Store) -> Result<SourceSet, EmitterError> {
        let mut sources = vec![];

        for attribute in store.attributes() {
            let mut fields = vec![];
            let subject = attribute.name.to_uri();
            if let Some(rels) = store.relations_by_subject.get(&subject) {
                for rel in rels {
                    let field_name = CamlFieldName::from_name(&rel.predicate);
                    let module_name = CamlModuleName::from_name(&rel.object);
                    let type_ref = CamlType::reference(module_name, "t".to_string());
                    fields.push((field_name, type_ref));
                }
            };

            let type_name = "t".to_string();
            let main_type = if fields.is_empty() {
                CamlType::abstract_type(type_name)
            } else {
                CamlType::record(type_name, fields)
            };

            let module = CamlModule::new(CamlModuleName::from_name(&attribute.name))
                .with_structure(vec![CamlValue::Type(main_type)]);
            sources.push(module.into())
        }

        for kind in store.kinds() {
            let mut fields = vec![];
            let subject = kind.name.to_uri();
            if let Some(rels) = store.relations_by_subject.get(&subject) {
                for rel in rels {
                    let field_name = CamlFieldName::from_name(&rel.predicate);
                    let module_name = CamlModuleName::from_name(&rel.object);
                    let type_ref = CamlType::reference(module_name, "t".to_string());
                    fields.push((field_name, type_ref));
                }
            };

            let type_name = "t".to_string();
            let main_type = if fields.is_empty() {
                CamlType::abstract_type(type_name)
            } else {
                CamlType::record(type_name, fields)
            };

            let module = CamlModule::new(CamlModuleName::from_name(&kind.name))
                .with_structure(vec![CamlValue::Type(main_type)]);
            sources.push(module.into())
        }

        Ok(SourceSet::from_sources(sources))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::*;

    macro_rules! test {
        ($name:ident, $src:expr) => {
            #[test]
            fn $name() {
                let mut store = lore_store::Store::new();
                let store = store.add_from_string($src).unwrap();
                let emitter = OCamlEmitter::new();
                let caml_value = emitter.translate(&store).unwrap();
                let snapshot = format!(
                    r#"
input:
    {}

output:

{:#?}
"#,
                    $src, caml_value
                );
                assert_snapshot!(snapshot)
            }
        };
    }

    test!(
        kind_to_type,
        r#"

prefix lore:rel:v1 as @lore

using dota:ontology:2022

kind Hero

attr Name

rel Hero @lore/hasOne Name

        "#
    );
}
