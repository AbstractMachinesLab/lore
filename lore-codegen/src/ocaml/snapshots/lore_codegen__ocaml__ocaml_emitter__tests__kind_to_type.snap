---
source: lore-codegen/src/ocaml/ocaml_emitter.rs
expression: snapshot

---

input:
    

prefix lore:rel:v1 as @lore

using dota:ontology:2022

kind Hero

attr Name

rel Hero @lore/hasOne Name

        

output:

SourceSet {
    sources: [
        Source {
            name: "dota_ontology_2022_name.mli",
            contents: "type t\n",
        },
        Source {
            name: "dota_ontology_2022_hero.mli",
            contents: "type t = { lore_rel_v1_hasOne: Dota_ontology_2022_name.t } \n\n",
        },
    ],
}

